extended/src/main/java/io/kubernetes/client/extended/controller/reconciler/Reconciler.java

```diff
@@ -0,0 +1,203 @@
+ * sources, etc).
+ *
+ * <p>Example reconcile Logic:
+ *
+ * <p>1. Read an object and all the Pods it owns. 2. Observe that the object spec specifies 5
+ * replicas but actual cluster contains only 1 Pod replica. 3. Create 4 Pods and set their
+ * OwnerReferences to the object.
+ *
+ * <p>Reconciliation is level-based, meaning action isn't driven off changes in individual Events,
+ * but instead is driven by actual cluster state read from the apiserver or a local cache. For
+ * example if responding to a Pod Delete Event, the Request won't contain that a Pod was deleted,
+ * instead the reconcile function observes this when reading the cluster state and seeing the Pod as
+ * missing.
+ */
+public interface Reconciler {
+
+  /**
+   * Reconcile result.
+   *
+   * @param request the reconcile request, triggered by watch events
+   * @return the result
+   */
+  Result reconcile(Request request);
+
+  /**
+   * The type Request contains the information necessary to reconcile a resource object. This
+   * includes the information to uniquely identify the object - its Name and Namespace. Note that it
+   * does NOT contain information about any specific Event or the object contents itself.
+   */
+  class Request {
+
+    /**
+     * Instantiates a new Request specifying the name. Works for cluster-scoped resource.
+     *
+     * @param name the resource object name
+     */
+    public Request(String name) {
+      this(null, name);
+    }
+
+    /**
+     * Instantiates a new Request specifying the name and namespace. Works for namespace-scoped
+     * resource.
+     *
+     * @param namespace the resource object namespace
+     * @param name the resource object name
+     */
+    public Request(String namespace, String name) {
+      this.name = name;
+      this.namespace = namespace;
+    }
+
+    private String name;
+    private String namespace;
+
+    /**
+     * Gets namespace.
+     *
+     * @return the namespace
+     */
+    public String getNamespace() {
+      return namespace;
+    }
+
+    /**
+     * Sets namespace.
+     *
+     * @param namespace the namespace
+     */
+    public void setNamespace(String namespace) {
+      this.namespace = namespace;
+    }
+
+    /**
+     * Gets name.
+     *
+     * @return the name
+     */
+    public String getName() {
+      return name;
+    }
+
+    @Override
+    public String toString() {
+      return "Request{" + "name='" + name + '\'' + ", namespace='" + namespace + '\'' + '}';
+    }
+
+    /**
+     * Sets name.
+     *
+     * @param name the name
+     */
+    public void setName(String name) {
+      this.name = name;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (this == o) return true;
+      if (o == null || getClass() != o.getClass()) return false;
+      Request request = (Request) o;
+      return Objects.equal(name, request.name) && Objects.equal(namespace, request.namespace);
```

Comment:
Do you need to use `Objects.equal(...)` here? can't you just do `name.equals(request.name)`?

Reply:
it's generated by IDE tooling, the two actually works the same i believe

Reply:
I think there definitely the same, but `Objects.equal(...)` feels unnecessarily verbose/hard to read....

Reply:
updated. refactored to old-school `equals` and `hashCode`